lib/WUI/link_content/basic_gets.cpp

```diff
@@ -0,0 +1,223 @@
+    char hostname[hostname_in_len + 1];
+    memcpy(hostname, hostname_unsynchronized, hostname_in_len);
+    hostname[hostname_in_len] = '\0';
+
+    // Keep the indentation of the JSON in here!
+    // clang-format off
+    JSON_START;
+    JSON_OBJ_START;
+        JSON_FIELD_STR("api", PL_VERSION_STRING) JSON_COMMA;
+        // FIXME: The server version is probably bogus. But it's unclear what
+        // the version is supposed to mean anyway. Waiting for the new
+        // PrusaLink API to replace this?
+        JSON_FIELD_STR("server", LWIP_VERSION_STRING) JSON_COMMA;
+        JSON_FIELD_STR("text", "PrusaLink MINI") JSON_COMMA;
+        JSON_FIELD_STR("hostname", hostname);
+    JSON_OBJ_END;
+    JSON_END;
+    // clang-format on
+}
+
+JsonRenderer::ContentResult get_job(size_t resume_point, JsonRenderer::Output &output) {
+    // Note about the marlin vars: It's true that on resumption we may get
+    // different values. But they would still be reasonably "sane". If we eg.
+    // finish a print and base what we include on previous version, we may
+    // outdated values, but they are still there.
+    marlin_vars_t *vars = marlin_vars();
+    marlin_client_loop();
+
+    bool has_job = false;
+    const char *state = "Unknown";
+
+    switch (vars->print_state) {
+    case mpsFinishing_WaitIdle:
+    case mpsFinishing_ParkHead:
+    case mpsPrinting:
+        has_job = true;
+        state = "Printing";
+        break;
+    case mpsPausing_Begin:
+    case mpsPausing_WaitIdle:
+    case mpsPausing_ParkHead:
+        has_job = true;
+        state = "Pausing";
+        break;
+    case mpsPaused:
+        has_job = true;
+        state = "Paused";
+        break;
+    case mpsResuming_Begin:
+    case mpsResuming_Reheating:
+    case mpsResuming_UnparkHead:
+        has_job = true;
+        state = "Resuming";
+        break;
+    case mpsAborting_Begin:
+    case mpsAborting_WaitIdle:
+    case mpsAborting_ParkHead:
+        has_job = true;
+        state = "Cancelling";
+        break;
+    case mpsAborted:
+    case mpsFinished:
+    case mpsIdle:
+        state = "Operational";
+        break;
+    }
+
+    // Keep the indentation of the JSON in here!
+    // clang-format off
+    JSON_START;
+    JSON_OBJ_START;
+        JSON_FIELD_STR("state", state) JSON_COMMA;
+        if (has_job) {
+            JSON_FIELD_OBJ("job");
+                JSON_FIELD_INT("estimatedPrintTime", vars->print_duration + vars->time_to_end) JSON_COMMA;
+                JSON_FIELD_OBJ("file")
+                    JSON_FIELD_STR("name", vars->media_LFN) JSON_COMMA;
+                    JSON_FIELD_STR("path", vars->media_SFN_path) JSON_COMMA;
+                    JSON_FIELD_STR("display", vars->media_LFN);
+                JSON_OBJ_END;
+            JSON_OBJ_END JSON_COMMA;
+            JSON_FIELD_OBJ("progress");
+                JSON_FIELD_FFIXED("completion", ((double)vars->sd_percent_done / 100.0)) JSON_COMMA;
```

Comment:
Division should be done in float, the result can be converted. Double is not supported by HW so it is emulated. Also multiplication by 0.01f will be faster. Not sure if compiler will convert it.

Reply:
OK with the float, though I believe it's minor (this code's speed is dominated by other factors, including all the sending of packets over network).

The compiler probably replaces the whole division by bitshifts or some such. Doing these things manually can actually throw it off a bit, so it's better to just write the code with the human reader in mind, the compiler will do a better job than us in this low-level stuff.